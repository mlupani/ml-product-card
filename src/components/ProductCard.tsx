import React, { createContext } from 'react';
import useProduct from '../hooks/useProduct';
import styles from '../styles/styles.module.css';
import { InitialValues, onChangeArgs, ProductCardContext, Producto, ProductCardHandlers } from '../interfaces/ProductCard';

export const ProductContext = createContext({} as ProductCardContext);
const { Provider } = ProductContext;

export interface Props {
    //children: React.ReactElement | React.ReactElement[];
    children: (args: ProductCardHandlers) => JSX.Element
    producto: Producto,
    className?: string,
    style?: React.CSSProperties,
    onChange?: (args: onChangeArgs) => void,
    value?: number
    initialValues?: InitialValues
}

export const ProductCard = ({ children, producto, className, style, onChange, value, initialValues }: Props) => {

    const { counter, increaseBy, reset, isMaxReached } = useProduct({ producto, onChange, value, initialValues });

    return (
        <Provider value={{ counter, increaseBy, producto, initialValues }}>
            <div className={`${styles.productCard} ${className}`} style={style}>
                {
                    children({
                        producto,
                        count: counter,
                        maxCount: initialValues?.maxCount,
                        increaseBy,
                        reset,
                        isMaxReached
                    })
                }
            </div>
        </Provider>
    )
}

export default ProductCard
