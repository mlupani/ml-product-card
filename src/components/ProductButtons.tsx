import React, { useCallback, useContext } from "react"
import { ProductContext } from "./ProductCard"
import styles from '../styles/styles.module.css';

export interface Props {
    className?: string,
    style?: React.CSSProperties
}

export const ProductButtons = ({className, style}: Props) => {

    const { increaseBy, counter, initialValues } = useContext(ProductContext)
    const isMaxReached = useCallback(
      () => !!initialValues?.maxCount && counter === initialValues.maxCount,
      [counter, initialValues?.maxCount],
    )

    return (
        <div className={`${styles.buttonsContainer} ${className}`} style={style}>
            <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>-</button>
            <div className={styles.countLabel}>{counter}</div>
            <button disabled={isMaxReached() ? true : false} className={`${styles.buttonAdd} ${isMaxReached() ? styles.disabled : null}`} onClick={() => increaseBy(+1)}>+</button>
        </div>
    )
}