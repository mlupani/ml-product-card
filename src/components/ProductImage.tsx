import React, { useContext } from "react";
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export interface Props {
    img?: string;
    className?: string;
    style?: React.CSSProperties;
}

export const ProductImage = ({img = '', className, style}: Props) => {

    const { producto } = useContext(ProductContext);
    let imageToShow;

    if (img) {
        imageToShow = img;
    } else if (producto.img) {
        imageToShow = producto.img;
    } else {
        imageToShow = noImage;
    }

    return (
        <img className={`${styles.productImg} ${className}`} src={imageToShow} alt="Product" style={style} />
    )
}