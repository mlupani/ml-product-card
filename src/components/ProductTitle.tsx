import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';

export interface Props {
    className?: string;
    title?: string;
    style?: React.CSSProperties;
}

export const ProductTitle = ({title, className= '', style} : Props) => {

    const { producto } = useContext(ProductContext);
    let titleToShow = title ? title : producto.title;

    return (
        <span className={`${styles.productDescription} ${className}`} style={style}>{titleToShow}</span>
    )
}