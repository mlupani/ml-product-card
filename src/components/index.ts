import { ProductTitle } from './ProductTitle';
import { ProductCard as ProductCardHOC } from './ProductCard';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';
import { ProductCardHOCProps } from '../interfaces/ProductCard';

export { ProductButtons } from "./ProductButtons";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";

export const ProductCard:ProductCardHOCProps = Object.assign(ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
});