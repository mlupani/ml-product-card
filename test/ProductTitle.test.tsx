import React from 'react';
import renderer from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../src/components';
import { product1 } from './data/products';

describe('productTitle', () => {
  it('renders the UI as expected', () => {
    const wrapper = renderer.create(<ProductTitle title={'Custom Product'} className={'Custom-class'} />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it('render productCard with product title', () => {
    const wrapper = renderer.create(<ProductCard producto={product1}>{() => <ProductTitle/>}</ProductCard>).toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});
