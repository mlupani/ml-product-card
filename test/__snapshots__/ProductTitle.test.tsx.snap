// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`productTitle render productCard with product title 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription "
  >
    Coffe mug - sin imagen
  </span>
</div>
`;

exports[`productTitle renders the UI as expected 1`] = `
<span
  className="productDescription Custom-class"
>
  Custom Product
</span>
`;
